x/auth/types/codec.go

```diff
@@ -5,24 +5,89 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/auth/exported"
 )
 
-// ModuleCdc auth module wide codec
-var ModuleCdc = codec.New()
+var _ AuthCodec = (*Codec)(nil)
+
+// AuthCodec defines the interface needed to serialize x/auth state. It must be
+// aware of all concrete account types.
+type AuthCodec interface {
+	codec.Marshaler
+
+	MarshalAccount(acc exported.AccountI) ([]byte, error)
+	UnmarshalAccount(bz []byte) (exported.AccountI, error)
+
+	MarshalAccountJSON(acc exported.AccountI) ([]byte, error)
+	UnmarshalAccountJSON(bz []byte) (exported.AccountI, error)
+}
+
+type Codec struct {
+	codec.Marshaler
+
+	// Keep reference to the amino codec to allow backwards compatibility along
+	// with type, and interface registration.
+	amino *codec.Codec
+}
+
+func NewCodec(amino *codec.Codec) *Codec {
+	return &Codec{Marshaler: codec.NewBaseCodec(amino), amino: amino}
+}
+
+// MarshalAccount marshals an AccountI interface. If the given type implements
+// the Marshaler interface, it is treated as a Proto-defined message and
+// serialized that way. Otherwise, it falls back on the internal Amino codec.
+func (c *Codec) MarshalAccount(acc exported.AccountI) ([]byte, error) {
+	return c.Marshaler.MarshalBinaryLengthPrefixed(acc)
```

Comment:
This isn't going to work because the `AccountI` types don't know how to get themselves back inside the `Account` `oneof`. You need to call `Account.SetAccountI` somehow. The overloading you want to allow is on the specific `Account` `oneof` and I don't see an easy way to do that with this approach. I think you're going to end up with something like https://github.com/cosmos/cosmos-sdk/pull/5474/files#diff-a26b2ab0f7188526d2f4e32346a37b1fR43 one way or another.

Reply:
Ahh, yes, you're correct. I've updated `MarshalAccount` to reflect this. Marshalling and unmarshalling `AccountI` works now.

Do you still see a problem with this approach?

Reply:
Well how would a user of `auth` support a different Account `oneof`? The approach I was taking allows that to be overloaded by passing in a factory function to the keeper. This approach seems to force the usage of fixed set of Account types.

Reply:
> Well how would a user of auth support a different Account oneof?

Exactly by passing a different `AuthCodec` implementation.

e.g.

The Hub uses vesting accounts, so it would not use `/x/auth/types/codec.go`, it would use `/x/auth/types/vesting/codec.go`. This codec implementation of `AuthCodec` has it's own `Account` with its own `oneof`s.